datarobot_batch_scoring/network.py

```diff
@@ -45,15 +46,23 @@ def _request(self, request):
                 callback = request.hooks['response'][0]
             response = FakeResponse(400, 'No Response')
             callback(response)
-        finally:
-            request.data = None
-            prepared.data = None
-            del request
-            del prepared
-            gc.collect()
 
     def perform_requests(self, requests):
-        return self._executor.map(self._request, requests)
+        for r in requests:
+            while True:
+                self.futures = [i for i in self.futures if not i.done()]
+                if len(self.futures) < self.concurrency:
+                    self.futures.append(self._executor.submit(self._request,
```

Comment:
We are iterating the WorkUnitGenerator sequentially. The batches generated by the WorkUnitGenerator are then being passed to the executor where they are being executed in threads non-sequentially. The only difference between this approach and the previous approach is that we do not submit additional batches until some of the outstanding threads finish executing. 
https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.submit